pulsar-transformations/src/main/java/com/datastax/oss/pulsar/functions/transforms/ComputeAIEmbeddingsStep.java

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.pulsar.functions.transforms;
+
+import com.datastax.oss.pulsar.functions.transforms.embeddings.EmbeddingsService;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+import lombok.Builder;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericData;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.pulsar.common.schema.SchemaType;
+
+/** Compute AI Embeddings for one or more records fields and put the value into a new or existing field. */
+@Builder
+public class ComputeAIEmbeddingsStep implements TransformStep {
+
+  @Builder.Default private final List<String> fields = new ArrayList<>();
+  @Builder.Default private final String embeddingsFieldName;
+  @Builder.Default private final EmbeddingsService embeddingsService;
+  private final Map<org.apache.avro.Schema, org.apache.avro.Schema> valueSchemaCache =
+      new ConcurrentHashMap<>();
+
+  @Override
+  public void process(TransformContext transformContext) {
+    if (!isSchemaCompatible(transformContext)) {
+      return;
+    }
+
+    Map<String, String> values = new HashMap<>();
+    collectFieldValuesFromKey(fields, transformContext, values);
+    collectFieldValuesFromValue(fields, transformContext, values);
+    if (values.isEmpty()) {
+      transformContext
+          .getContext()
+          .getLogger()
+          .debug("Skipping message, empty values for configured fields");
+      return;
+    }
+    final String text =
+        fields
+            .stream()
+            .map(f -> values.get(f))
+            .filter(Objects::nonNull)
+            .collect(Collectors.joining(" "));
+
+    final List<Double> embeddings = embeddingsService.computeEmbeddings(List.of(text)).get(0);
+    applyEmbeddingsToRecord(transformContext, embeddings);
+  }
+
+  private void applyEmbeddingsToRecord(TransformContext transformContext, List<Double> embeddings) {
```

Comment:
I think the first release could be to output the value with Schema.STRING

Reply:
it's a compute transformation so IMHO the expectation is to have the same message enriched with the embeddings. 
Actually, in case of primitive schemas, we could output the embeddings instead of the value.